core/src/Instance/B2CAuthority.cs

```diff
@@ -37,37 +38,85 @@ internal class B2CAuthority : AadAuthority
     {
         public const string Prefix = "tfp"; // The http path of B2C authority looks like "/tfp/<your_tenant_name>/..."
         public const string B2CCanonicalAuthorityTemplate = "https://{0}/{1}/{2}/{3}/";
+        public const string MicrosoftOnline = "https://login.microsoftonline.com";
+        public const string OpenIdConfigurationEndpoint = "v2.0/.well-known/openid-configuration";
+        internal const string B2CTrustedHost = "b2clogin.com";
 
-        internal B2CAuthority(string authority, bool validateAuthority) 
+        internal B2CAuthority(string authority, bool validateAuthority)
             : base(authority, validateAuthority)
         {
+            AuthorityType = AuthorityType.B2C;
+            ValidateAuthority = validateAuthority;
+
             Uri authorityUri = new Uri(authority);
-            string[] pathSegments = authorityUri.AbsolutePath.Substring(1).Split(new [] { '/'}, StringSplitOptions.RemoveEmptyEntries);
+            string[] pathSegments = authorityUri.AbsolutePath.Substring(1).Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
             if (pathSegments.Length < 3)
             {
                 throw new ArgumentException(CoreErrorMessages.B2cAuthorityUriInvalidPath);
             }
 
             CanonicalAuthority = string.Format(CultureInfo.InvariantCulture, B2CCanonicalAuthorityTemplate, authorityUri.Authority,
                 pathSegments[0], pathSegments[1], pathSegments[2]);
-
-            AuthorityType = AuthorityType.B2C;
         }
 
-        protected override async Task<string> GetOpenIdConfigurationEndpointAsync(
-            IHttpManager httpManager, 
-            ITelemetryManager telemetryManager, 
-            string userPrincipalName, 
+        internal override async Task UpdateCanonicalAuthorityAsync(
+            IHttpManager httpManager,
+            ITelemetryManager telemetryManager,
             RequestContext requestContext)
         {
-            if (ValidateAuthority && !IsInTrustedHostList(new Uri(CanonicalAuthority).Host))
+            if (ValidateAuthority && !InTrustedHostList(new Uri(CanonicalAuthority).Host))
             {
                 throw new ArgumentException(CoreErrorMessages.UnsupportedAuthorityValidation);
             }
+            
+            var metadata = await AadInstanceDiscovery
+                             .Instance.GetMetadataEntryAsync(
+                                 httpManager,
+                                 telemetryManager,
+                                 new Uri(UpdateCanonicalAuthorityB2C()),
+                                 ValidateAuthority,
+                                 requestContext)
+                             .ConfigureAwait(false);
+
+            CanonicalAuthority = UpdateHost(CanonicalAuthority, metadata.PreferredNetwork);
+        }
+
+        private string UpdateCanonicalAuthorityB2C()
```

Comment:
This method name does not match what it is doing. This is a "query" type of method, i.e. it returns some info without changing any state, but you named it as a "command" method, i.e. it changes state. 

Also, I think this can be static since it only uses some constants from this class, i.e. it doesn't use state.

Reply:
what name would you suggest? it is changing the value of the canonical authority, because we need to replace the host w/microsoftonline.com for the b2clogin.com scenarios.

---
In reply to: [230920739](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1383#discussion_r230920739) [](ancestors = 230920739)

Reply:
GetCannonicalAuthorityB2C

---
In reply to: [230939266](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1383#discussion_r230939266) [](ancestors = 230939266,230920739)

Reply:
i thought of that name too, but wasn't sure, so i'll go w/it since you suggested it too. :)

---
In reply to: [231095905](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1383#discussion_r231095905) [](ancestors = 231095905,230939266,230920739)